README.md
# Ruby API library for Doxie Go Wifi

[![Gem Version](https://badge.fury.io/rb/doxie.svg)](https://badge.fury.io/rb/doxie) [![Build Status](https://travis-ci.com/cbetta/doxie.svg?branch=master)](https://travis-ci.com/cbetta/doxie)

A client library for the API available on WiFi-enabled [Doxie scanners](http://getdoxie.com). The specification of the API is available in their [developer documentation](/docs).

## Scanners Supported

This library supports the following scanners:

* Doxie Go Wifi (DX250)
* Doxie Go SE (DX255)
* Doxie Q (DX300)

## Installation

Either install directly or via bundler.

```rb
gem 'doxie'
gem 'doxie-scanner' # optional if your Doxie is not on a fixed IP
```

## Usage

### Optional: Finding your Doxie's IP

This optional step requires the [`doxie-scanner`](https://github.com/cbetta/doxie-scanner) gem. This gem has a bigger dependency than the `doxie` gem, which is why it has been intentionally separated into a seperate library.

```rb
require 'doxie/scanner'
Doxie::Scanner.ips
=> [
    [0] "192.168.1.2"
]
```

### Client

The client accepts an `ip` address, a scanner model number (defaults to `Doxie::GO`) and an optional `password`. You can omit the `password` if your Doxie has non set.

```rb
require 'doxie'
client = Doxie::Client.new(
    ip: '192.168.1.2', 
    model: Doxie::Q, # of Doxie::GO, or Doxie::GO_SE
    password: 'test'
)
```

Alternatively, you can use the [`doxie-scanner`](https://github.com/cbetta/doxie-scanner) gem to automatically connect to any scanner found. 

```rb
require 'doxie'
require 'doxie/scanner'

client = Doxie::Client.new(
  Doxie::Scanner.devices.first
)
```

### `GET /hello.json`

Returns status information for the scanner, firmware, network mode, and password
configuration. Accessing this command does not require a password if one has been
set. The values returned depend on whether the scanner is creating its own
network or joining an existing network.

```rb
client.hello
=> {
           "model" => "DX250",
            "name" => "Doxie_062300",
    "firmwareWiFi" => "1.29",
     "hasPassword" => true,
             "MAC" => "00:11:11:11:11:00",
            "mode" => "Client",
         "network" => "YourWifi",
              "ip" => "192.168.1.2"
}
```

* `model`: DX250 for the Doxie Go WiFi, DX255 for the Doxie Go SE, and DX300 for the Doxie Q.
* `name`: The name of the scanner, which defaults to the form "Doxie_XXXXXX".
  The name of a scanner can be changed by using the Doxie desktop app.
* `firmwareWiFi`: The Wi-Fi firmware version.
* `hasPassword`: Indicates whether a password has been set to authenticate API
  access. Passwords can be set and removed by using the Doxie desktop app.
* `MAC`: The MAC address of the scanner as shown on the scanner's bottom
  label.
* `mode`: "AP" if the scanner is creating its own network or "Client" if the
  scanner is joining an existing network.
* `network`: If the scanner is in "Client" mode, this is the name of the
  network it has joined.
* `ip`: If the scanner is in "Client" mode, this is the IP of the scanner on
  the network it has joined.

### `GET /hello_extra.json`

Returns additional status values. These values are accessed separately from
those in `/hello.json` because there can be a delay of several seconds in
loading them. Accessing this command does not require a password if one has
been set.

```rb
client.hello_extra
=> {
                    "firmware" => "0.26",
    "connectedToExternalPower" => true
}
```

* `firmware`: The scanner firmware version.
* `connectedToExternalPower`: Indicates whether the scanner is connected to
  its AC adapter versus running on battery power. This value is not cached, so
  it immediately reflects any state changes.

This method is only available for the original Doxie Go WiFi model.

### `GET /restart.json`

Restarts the scanner's Wi-Fi system. The scanner's status light blinks blue
during the restart.

```rb
client.restart
=> true
```

### `GET /scans.json`

Returns an array of all scans currently in the scanner’s memory. After scanning
a document, the scan will available via the API several second later.

```rb
client.scans
=> [
    [0] {
            "name" => "/DOXIE/JPEG/IMG_0001.JPG",
            "size" => 900964,
        "modified" => "2010-05-01 00:02:38"
    }
]
```

Calling this function immediately after scanning something may return a blank
result, even if there are other scans on the scanner, due to the scanner's
memory being in use. Consider retrying if a successful HTTP status code is
returned along with a blank body.

### `GET /scans/recent.json`

Returns the path to the last scan if available. Monitoring this value for
changes provides a simple way to detect new scans without having to fetch the
entire list of scans.

```rb
client.recent_scans
=> {
    "path" => "/DOXIE/JPEG/IMG_0001.JPG"
}
```

### `GET /scans/DOXIE/JPEG/IMG_XXXX.JPG`

There are 2 ways to get a scan off your Doxie. The first is to get the raw binary content and then do something with it yourself.

```rb
client.scan "/DOXIE/JPEG/IMG_0001.JPG"
=> "...?]?1:Xt?????'A??}:<??13???z*???}?rT???????z!ESj?/?..."
```

The other is to pass in a filename:

```rb
client.scan "/DOXIE/JPEG/IMG_0001.JPG", 'test.jpg'
=> true
```

### `GET /thumbnails/DOXIE/JPEG/IMG_XXXX.JPG`

There are 2 ways to get a thumbnail off your Doxie. The first is to get the raw binary content and then do something with it yourself.

```rb
client.thumbnail "/DOXIE/JPEG/IMG_0001.JPG"
=> "...?]?1:Xt?????'A??}:<??13???z*???}?rT???????z!ESj?/?..."
```

The other is to pass in a filename:

```rb
client.thumbnail "/DOXIE/JPEG/IMG_0001.JPG", 'test.jpg'
=> true
```

Thumbnails are constrained to fit within 240x240 pixels. Thumbnails for new
scans are not generated until after the scan has been made available in
`/scans.json` and `/scans/recent.json`. This function will return 404 Not Found
if the thumbnail has not yet been generated. Retrying after a delay is
recommended to handle such cases.

### `DELETE /scans/DOXIE/JPEG/IMG_XXXX.JPG`

Deletes the scan at the specified path.

```rb
client.delete_scan "/DOXIE/JPEG/IMG_0001.JPG"
=> true
```

Deleting takes several seconds because a lock on the internal storage must be
obtained and released. Deleting may fail if the lock cannot be obtained
(e.g., the scanner is busy), so consider retrying on failure conditions. When
deleting multiple scans, use `/scans/delete.json` for best performance.

This will raise an error if the file is no longer present (`Doxie::Client::Error<Net::HTTPForbidden>`)

### `POST /scans/delete.json`

Deletes multiple scans in a single operation. This is much faster than deleting
each scan individually.

```rb
client.delete_scans ["/DOXIE/JPEG/IMG_0001.JPG", "/DOXIE/JPEG/IMG_0002.JPG"]
=> true
```

This will raise an error if the files are no longer present (`Doxie::Client::Error<Net::HTTPForbidden>`)

## Contributing

 1. **Fork** the repo on GitHub
 2. **Clone** the project to your own machine
 3. **Commit** changes to your own branch
 4. **Push** your work back up to your fork
 5. Submit a **Pull request** so that we can review your changes

### Development

* `bundle install` to get dependencies
* `rake` to run tests
* `rake console` to run a local console with the library loaded

## Credits and License

Thanks to [@timcraft](https://github.com/timcraft) for the excellent [Nexmo Ruby library](https://github.com/Nexmo/nexmo-ruby) which helped me remember how to nicely wrap the Doxie API.

This library is released under the [MIT License](LICENSE).


CHANGELOG
* 4.0.0 Remove support for Ruby 2.3/2.4
* 3.0.0 Add support for latest Doxie scanners
* 2.0.0 Updated gems, rubocop checks, and changed error response
* 1.0.0 Initial release


spec/doxie_spec.rb
# frozen_string_literal: true

require 'minitest/autorun'
require 'webmock/minitest'
require 'fakefs'
require 'doxie'

describe 'Doxie::Client' do
  def json_response_body(content)
    {
      headers: {
        'Content-Type' => 'application/json;charset=utf-8'
      },
      body: content
    }
  end

  before do
    @json_response_object = { 'key' => 'value' }
    @json_response_body = json_response_body('{"key":"value"}')
    @ip = '192.168.1.1'
    @base_url = "http://#{@ip}:8080"
    @base_url_v2 = "http://#{@ip}:80"
    @client = Doxie::Client.new(ip: @ip)
  end

  describe 'Doxie Models' do
    it 'should assign the right values to each model' do
      assert_equal Doxie::GO, Doxie::API_V1
      assert_equal Doxie::DX250, Doxie::API_V1

      assert_equal Doxie::Q, Doxie::API_V2
      assert_equal Doxie::GO_SE, Doxie::API_V2
      assert_equal Doxie::DX255, Doxie::API_V2
      assert_equal Doxie::DX300, Doxie::API_V2
    end
  end

  describe 'get /hello.json' do
    it 'should return the result' do
      stub_request(:get, "#{@base_url}/hello.json")
        .to_return(@json_response_body)
      _(@client.hello).must_equal(@json_response_object)
    end
  end

  describe 'get /hello_extra.json' do
    it 'should return the result' do
      stub_request(:get, "#{@base_url}/hello_extra.json")
        .to_return(@json_response_body)
      _(@client.hello_extra).must_equal(@json_response_object)
    end

    it 'should error for API V2 models, as the method does not exist' do
      @client = Doxie::Client.new(ip: @ip, model: Doxie::API_V2)
      error = _(-> { @client.hello_extra }).must_raise(Doxie::Client::Error)
      _(error.message).must_match 'Method does not exist for this model'
    end
  end

  describe 'get /restart.json' do
    it 'should return the result' do
      stub_request(:get, "#{@base_url}/restart.json")
        .to_return(status: 204)
      _(@client.restart).must_equal(true)
    end
  end

  describe 'get /scans.json' do
    it 'should return the result' do
      stub_request(:get, "#{@base_url}/scans.json")
        .to_return(@json_response_body)
      _(@client.scans).must_equal(@json_response_object)
    end

    it 'should return an empty array when there are no scans on a V2 model' do
      @client = Doxie::Client.new(ip: @ip, model: Doxie::API_V2)
      stub_request(:get, "#{@base_url_v2}/scans.json")
        .to_return(status: 404)
      _(@client.scans).must_equal([])
    end
  end

  describe 'get /scans/recent.json' do
    it 'should return the result' do
      stub_request(:get, "#{@base_url}/scans/recent.json")
        .to_return(@json_response_body)
      _(@client.recent_scans).must_equal(@json_response_object)
    end

    it 'should return an empty array when there are no scans on a V2 model' do
      @client = Doxie::Client.new(ip: @ip, model: Doxie::API_V2)
      stub_request(:get, "#{@base_url_v2}/scans/recent.json")
        .to_return(status: 204)
      _(@client.recent_scans).must_equal([])
    end
  end

  describe 'get /scans/DOXIE/JPEG/IMG_0001.JPG' do
    it 'should return the result' do
      stub_request(:get, "#{@base_url}/scans/DOXIE/JPEG/IMG_0001.JPG")
        .to_return(@json_response_body)
      _(@client.scan('/DOXIE/JPEG/IMG_0001.JPG'))
        .must_equal(@json_response_object)
    end

    it 'should write to file' do
      stub_request(:get, "#{@base_url}/scans/DOXIE/JPEG/IMG_0001.JPG")
        .to_return(@json_response_body)
      _(@client.scan('/DOXIE/JPEG/IMG_0001.JPG', 'test.jpg')).must_equal(true)
    end
  end

  describe 'get /thumbnails/DOXIE/JPEG/IMG_0001.JPG' do
    it 'should return the result' do
      stub_request(:get, "#{@base_url}/thumbnails/DOXIE/JPEG/IMG_0001.JPG")
        .to_return(@json_response_body)
      _(@client.thumbnail('/DOXIE/JPEG/IMG_0001.JPG'))
        .must_equal(@json_response_object)
    end

    it 'should write to file' do
      stub_request(:get, "#{@base_url}/thumbnails/DOXIE/JPEG/IMG_0001.JPG")
        .to_return(@json_response_body)
      _(@client.thumbnail('/DOXIE/JPEG/IMG_0001.JPG', 'test.jpg'))
        .must_equal(true)
    end
  end

  describe 'DELETE /scans/DOXIE/JPEG/IMG_0001.JPG' do
    it 'should return the result' do
      stub_request(:delete, "#{@base_url}/scans/DOXIE/JPEG/IMG_0001.JPG")
        .to_return(@json_response_body)
      _(@client.delete_scan('/DOXIE/JPEG/IMG_0001.JPG'))
        .must_equal(@json_response_object)
    end
  end

  describe 'POST /scans/delete.json' do
    it 'should return the result' do
      stub_request(:post, "#{@base_url}/scans/delete.json")
        .to_return(status: 204)
      _(@client.delete_scans(['/DOXIE/JPEG/IMG_0001.JPG']))
        .must_equal(true)
    end
  end

  it 'raises an authentication error exception if the response code is 401' do
    stub_request(:get, "#{@base_url}/hello.json")
      .to_return(status: 401)
    _(proc { @client.hello }).must_raise(Doxie::Client::Error)
  end

  it 'raises a client error exception if the response code is 4xx' do
    stub_request(:get, "#{@base_url}/hello.json")
      .to_return(status: 400)
    _(proc { @client.hello }).must_raise(Doxie::Client::Error)
  end

  it 'raises a server error exception if the response code is 5xx' do
    stub_request(:get, "#{@base_url}/hello.json")
      .to_return(status: 500)
    _(proc { @client.hello }).must_raise(Doxie::Client::Error)
  end
end


lib/doxie/client.rb
# frozen_string_literal: true

require 'net/http'
require 'json'

module Doxie
  # The client for connecting to a Doxie scanner.
  #
  # Use the IP and password to connect as follows:
  #   Doxie::Client.new(ip: '192.168.1.2', model: Doxie::Q, password: 'test')
  class Client
    class Error < StandardError; end

    USERNAME = 'doxie'

    attr_accessor :ip, :password, :model, :port

    def initialize(options)
      @ip = options[:ip] || ''
      @password = options[:password] || ''
      @model = options[:model] || Doxie::API_V1
      @port = @model == Doxie::API_V1 ? 8080 : 80
    end

    def hello
      get('/hello.json')
    end

    def hello_extra
      if model == Doxie::API_V2
        raise Doxie::Client::Error, 'Method does not exist for this model'
      end

      get('/hello_extra.json')
    end

    def restart
      get('/restart.json') || true
    end

    def scans
      get('/scans.json')
    rescue Doxie::Client::Error => e
      # a 404 is thrown on the Doxie Q and
      # Doxie GO SE when there are no scans
      raise e if model == Doxie::API_V1

      []
    end

    def recent_scans
      get('/scans/recent.json') || []
    end

    def scan(scan_name, file_name = nil)
      file "/scans#{scan_name}", file_name
    end

    def thumbnail(scan_name, file_name = nil)
      file "/thumbnails#{scan_name}", file_name
    end

    def delete_scan(scan_name)
      delete("/scans#{scan_name}") || true
    end

    def delete_scans(scan_names)
      post('/scans/delete.json', scan_names) || true
    end

    private

    def get(path)
      uri = uri_for(path)
      message = Net::HTTP::Get.new(uri.request_uri)
      parse(request(uri, message))
    end

    def post(path, params)
      uri = uri_for(path)
      message = Net::HTTP::Post.new(uri.request_uri)
      message.body = JSON.generate(params)
      parse(request(uri, message))
    end

    def delete(path)
      uri = uri_for(path)
      message = Net::HTTP::Delete.new(uri.request_uri)
      parse(request(uri, message))
    end

    def uri_for(path)
      URI("http://#{ip}:#{port}#{path}")
    end

    def request(uri, message)
      message.basic_auth USERNAME, password if password && !password.empty?
      http = Net::HTTP.new(uri.host, uri.port)
      http.request(message)
    end

    def parse(response)
      case response
      when Net::HTTPNoContent
        nil
      when Net::HTTPSuccess
        parse_json(response)
      else
        raise Error, response
      end
    end

    def parse_json(response)
      if response['Content-Type'].split(';').first == 'application/json'
        JSON.parse(response.body)
      else
        response.body
      end
    end

    def file(scan_name, file_name)
      body = get(scan_name)
      if file_name
        File.open(file_name, 'wb') { |file| file.write(body) }
        true
      else
        body
      end
    end
  end
end
